pkg/vpcmodel/multiExplainability.go

```diff
@@ -0,0 +1,180 @@
+	for i, v := range srcDstCouples {
+		emptyExplain := &Explanation{nil, nil, nil, v.src.Name(), v.dst.Name(),
+			nil, nil, false, nil}
+		if v.c == nil {
+			// no vpc config implies missing cross-vpc router between src and dst which are not in the same VPC
+			multiExplanation[i] = explainOutputEntry{emptyExplain, nil}
+			continue
+		}
+		srcNodes, errSrc := getNodesFromEndpoint(v.src)
+		if errSrc != nil {
+			multiExplanation[i] = explainOutputEntry{emptyExplain, errSrc}
+			continue
+		}
+		dstNodes, errDst := getNodesFromEndpoint(v.dst)
+		if errDst != nil {
+			multiExplanation[i] = explainOutputEntry{emptyExplain, errDst}
+			continue
+		}
+		var connectivity *VPCConnectivity
+		var ok bool
+		if connectivity, ok = vpcConns[v.c.VPC.UID()]; !ok {
+			errConn := fmt.Errorf("npGuard eror: missing connectivity computation for %v %v in MultiExplain", v.c.VPC.UID(), v.c.VPC.Name())
+			multiExplanation[i] = explainOutputEntry{emptyExplain, errConn}
+			continue
+		}
+		explain, errExplain := v.c.explainConnectivityForVPC(v.src.Name(), v.dst.Name(), srcNodes, dstNodes,
+			srcDstInternalAddr{false, false}, nil, connectivity)
+		if errExplain != nil {
+			multiExplanation[i] = explainOutputEntry{emptyExplain, errExplain}
+			continue
+		}
+		multiExplanation[i] = explainOutputEntry{explain, nil}
+	}
+	return multiExplanation
+}
+
+// given an Endpoint, return []Node which is either:
+// 1. A single Node representing a VSI if the endpoints consists a single vsi
+// 2. A number of Nodes, each representing an external address, if the endpoint is groupedExternalNodes
+// if the endpoint is neither, returns error
+func getNodesFromEndpoint(endpoint EndpointElem) ([]Node, error) {
+	switch n := endpoint.(type) {
+	case InternalNodeIntf:
+		return []Node{endpoint.(Node)}, nil
+	case *groupedExternalNodes:
+		externalNodes := make([]Node, len(*n))
+		for i, e := range *n {
+			externalNodes[i] = e
+		}
+		return externalNodes, nil
+	}
+	return nil, fmt.Errorf("np-Guard error: %v not of type InternalNodeIntf or groupedExternalNodes", endpoint.Name())
+}
+
+// CreateMultiExplanationsInput() given configs and results of connectivity analysis, generates input
+// in the format required by MultiExplain
+// it creates the explainInputEntry of all the following cases:
+// (1) src and dst are internal nodes from the same vpc:                                 {src,dst,vpcConfig}
+// (2) src is internal and dst is external:                                              {src,dst,srcVpcConfig}
+// (3) dst is internal and src is external:                                              {src,dst,dstVpcConfig}
+// (4) src and dst are internal nodes from different vpcs and has a multivpc connection: {src,dst,multiVpcConfig}
+// (5) src and dst are internal nodes from different vpcs and has no connection:         {src,dst,nil}
+func CreateMultiExplanationsInput(cConfigs MultipleVPCConfigs, conns map[string]*GroupConnLines) []explainInputEntry {
```

Comment:
Seems this function has issues - see https://github.com/np-guard/vpc-network-config-analyzer/issues/535#issuecomment-2062976516

Reply:
better also add tests to this function separately.

Reply:
> Seems this function has issues - see [#535 (comment)](https://github.com/np-guard/vpc-network-config-analyzer/issues/535#issuecomment-2062976516)

The issue is not in this function, I believe. See discussion in #535

Reply:
issue fixed